File: src/bun.js/test/expect/toHaveReturned.zig
inline fn toHaveReturnedTimesFn(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame, comptime mode: enum { toHaveReturned, toHaveReturnedTimes }) bun.JSError!JSValue {
    jsc.markBinding(@src());

    const thisValue = callframe.this();
    const arguments = callframe.arguments();
    defer this.postMatch(globalThis);

    const value: JSValue = try this.getValue(globalThis, thisValue, @tagName(mode), "<green>expected<r>");

    this.incrementExpectCallCounter();

    var returns = try mock.jestMockIterator(globalThis, value);

    const expected_success_count: i32 = if (mode == .toHaveReturned) brk: {
        if (arguments.len > 0 and !arguments[0].isUndefined()) {
            return globalThis.throwInvalidArguments(@tagName(mode) ++ "() must not have an argument", .{});
        }
        break :brk 1;
    } else brk: {
        if (arguments.len < 1 or !arguments[0].isUInt32AsAnyInt()) {
            return globalThis.throwInvalidArguments(@tagName(mode) ++ "() requires 1 non-negative integer argument", .{});
        }

        break :brk try arguments[0].coerce(i32, globalThis);
    };

    var pass = false;

    var actual_success_count: i32 = 0;
    var total_call_count: i32 = 0;
    while (try returns.next()) |item| {
        switch (try mock.jestMockReturnObject_type(globalThis, item)) {
            .@"return" => actual_success_count += 1,
            else => {},
        }
        total_call_count += 1;
    }

    pass = switch (mode) {
        .toHaveReturned => actual_success_count >= expected_success_count,
        .toHaveReturnedTimes => actual_success_count == expected_success_count,
    };

    const not = this.flags.not;
    if (not) pass = !pass;
    if (pass) return .js_undefined;

    switch (not) {
        inline else => |is_not| {
            const signature = comptime getSignature(@tagName(mode), "<green>expected<r>", is_not);
            const str: []const u8, const spc: []const u8 = switch (mode) {
                .toHaveReturned => switch (not) {
                    false => .{ ">= ", "   " },
                    true => .{ "< ", "  " },
                },
                .toHaveReturnedTimes => switch (not) {
                    false => .{ "== ", "   " },
                    true => .{ "!= ", "   " },
                },
            };
            return this.throw(globalThis, signature,
                \\
                \\
                \\Expected number of succesful returns: {s}<green>{d}<r>
                \\Received number of succesful returns: {s}<red>{d}<r>
                \\Received number of calls:             {s}<red>{d}<r>
                \\
            , .{ str, expected_success_count, spc, actual_success_count, spc, total_call_count });
        },
    }
}

pub fn toHaveReturned(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame) bun.JSError!JSValue {
    return toHaveReturnedTimesFn(this, globalThis, callframe, .toHaveReturned);
}

pub fn toHaveReturnedTimes(this: *Expect, globalThis: *JSGlobalObject, callframe: *CallFrame) bun.JSError!JSValue {
    return toHaveReturnedTimesFn(this, globalThis, callframe, .toHaveReturnedTimes);
}

const bun = @import("bun");

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;
const mock = bun.jsc.Expect.mock;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/runtime/test_runner/expect/toHaveReturned.rs
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
#[allow(unused_imports)] use super::{JSValueTestExt, JSGlobalObjectTestExt, BigIntCompare, make_formatter};

use super::mock;
use super::Expect;

#[derive(PartialEq, Eq, Clone, Copy)]
enum Mode {
    ToHaveReturned,
    ToHaveReturnedTimes,
}

impl Mode {
    #[inline]
    const fn tag_name(self) -> &'static str {
        match self {
            Mode::ToHaveReturned => "toHaveReturned",
            Mode::ToHaveReturnedTimes => "toHaveReturnedTimes",
        }
    }
}

// PERF(port): Zig used a `comptime mode` parameter (anonymous enum) so the two callers were
// monomorphized; stable Rust forbids enum const-generic params (`adt_const_params`). Passed as a
// runtime value here — the body branches on it only on cold/error paths.
#[inline]
fn to_have_returned_times_fn(
    this: &Expect,
    global: &JSGlobalObject,
    callframe: &CallFrame,
    mode: Mode,
) -> JsResult<JSValue> {
    bun_jsc::mark_binding!();
    let arguments = callframe.arguments();
    let (this, returns_arr, _value) = this.mock_prologue(
        global,
        callframe.this(),
        mode.tag_name(),
        "<green>expected<r>",
        mock::MockKind::Returns,
    )?;
    let mut returns = returns_arr.array_iterator(global)?;

    let expected_success_count: i32 = if mode == Mode::ToHaveReturned {
        if arguments.len() > 0 && !arguments[0].is_undefined() {
            // PERF(port): Zig used comptime `@tagName(mode) ++ "..."`; runtime fmt on error path.
            return Err(global.throw_invalid_arguments(format_args!(
                "{}() must not have an argument",
                mode.tag_name()
            )));
        }
        1
    } else {
        if arguments.len() < 1 || !arguments[0].is_uint32_as_any_int() {
            return Err(global.throw_invalid_arguments(format_args!(
                "{}() requires 1 non-negative integer argument",
                mode.tag_name()
            )));
        }

        arguments[0].coerce::<i32>(global)?
    };

    let mut pass;

    let mut actual_success_count: i32 = 0;
    let mut total_call_count: i32 = 0;
    while let Some(item) = returns.next()? {
        match mock::jest_mock_return_object_type(global, item)? {
            mock::ReturnStatus::Return => actual_success_count += 1,
            _ => {}
        }
        total_call_count += 1;
    }

    pass = match mode {
        Mode::ToHaveReturned => actual_success_count >= expected_success_count,
        Mode::ToHaveReturnedTimes => actual_success_count == expected_success_count,
    };

    let not = this.flags.get().not();
    if not {
        pass = !pass;
    }
    if pass {
        return Ok(JSValue::UNDEFINED);
    }

    // Zig: `switch (not) { inline else => |is_not| ... }` — runtime bool → comptime dispatch.
    // PERF(port): Zig computed `getSignature` at comptime; runtime here (error path, cold).
    let signature = Expect::get_signature(mode.tag_name(), "<green>expected<r>", not);
    // `throw` runs pretty_fmt over the *rendered* string, so `<`/`>` in these
    // operands must be backslash-escaped to survive the tag pass.
    let (str_, spc): (&'static str, &'static str) = match mode {
        Mode::ToHaveReturned => match not {
            false => ("\\>= ", "   "),
            true => ("\\< ", "  "),
        },
        Mode::ToHaveReturnedTimes => match not {
            false => ("== ", "   "),
            true => ("!= ", "   "),
        },
    };
    this.throw(
        global,
        signature,
        format_args!(
            "\n\n\
             Expected number of succesful returns: {}<green>{}<r>\n\
             Received number of succesful returns: {}<red>{}<r>\n\
             Received number of calls:             {}<red>{}<r>\n",
            str_, expected_success_count, spc, actual_success_count, spc, total_call_count,
        ),
    )
}

impl Expect {
    #[bun_jsc::host_fn(method)]
    pub fn to_have_returned(
        &self,
        global: &JSGlobalObject,
        callframe: &CallFrame,
    ) -> JsResult<JSValue> {
        to_have_returned_times_fn(self, global, callframe, Mode::ToHaveReturned)
    }

    #[bun_jsc::host_fn(method)]
    pub fn to_have_returned_times(
        &self,
        global: &JSGlobalObject,
        callframe: &CallFrame,
    ) -> JsResult<JSValue> {
        to_have_returned_times_fn(self, global, callframe, Mode::ToHaveReturnedTimes)
    }
}

// ported from: src/test_runner/expect/toHaveReturned.zig
